widget/diagramwidget/diagram.go

```diff
@@ -0,0 +1,251 @@
+package diagramwidget
+
+import (
+	"image/color"
+
+	"fyne.io/fyne/v2"
+	"fyne.io/fyne/v2/canvas"
+	"fyne.io/fyne/v2/driver/desktop"
+	"fyne.io/fyne/v2/theme"
+	"fyne.io/fyne/v2/widget"
+)
+
+var Globaldiagram *DiagramWidget
+
+// ForceRepaint is a workaround for a Fyne bug (Issue #2205) in which moving a canvas object does not
+// trigger repainting. When the issue is resolved, this function and all references to it should be
+// removed. The DummyBox on the GlobalDiagram should also be removed.
+// The conditionals here are required during initialization.
+func ForceRepaint() {
+	if Globaldiagram != nil {
+		if Globaldiagram.DummyBox != nil {
+			Globaldiagram.DummyBox.Refresh()
+		}
+	}
+}
+
+// Verify that interfaces are fully implemented
+var _ fyne.Tappable = (*DiagramWidget)(nil)
+
+type linkPinPair struct {
+	link *DiagramLink
+	pin  *LinkPoint
+}
+
+type DiagramWidget struct {
+	widget.BaseWidget
+
+	// ID is expected to be unique across all DiagramWidgets in the application.
+	ID string
+
+	// Diagrams may want to use a different theme and variant than the application. The default value is the
+	// applicaton's theme and variant
+	DiagramTheme fyne.Theme
+	ThemeVariant fyne.ThemeVariant
+	Offset       fyne.Position
+
+	// DesiredSize specifies the size which the graph widget should take
+	// up, defaults to 800 x 600
+	DesiredSize fyne.Size
+
+	Nodes                          map[string]*DiagramNode
+	Links                          map[string]*DiagramLink
+	selection                      map[string]DiagramElement
+	diagramElementLinkDependencies map[string][]linkPinPair
+
+	// TODO Remove DummyBox when fyne rendering issue is resolved
+	DummyBox *canvas.Rectangle
+}
+
+func NewDiagramWidget(id string) *DiagramWidget {
+	dw := &DiagramWidget{
+		ID:                             id,
+		DiagramTheme:                   fyne.CurrentApp().Settings().Theme(),
+		ThemeVariant:                   fyne.CurrentApp().Settings().ThemeVariant(),
+		DesiredSize:                    fyne.Size{Width: 800, Height: 600},
+		Offset:                         fyne.Position{X: 0, Y: 0},
+		Nodes:                          map[string]*DiagramNode{},
+		Links:                          map[string]*DiagramLink{},
+		DummyBox:                       canvas.NewRectangle(color.Transparent),
+		selection:                      map[string]DiagramElement{},
+		diagramElementLinkDependencies: map[string][]linkPinPair{},
+	}
+	dw.DummyBox.SetMinSize(fyne.Size{Height: 50, Width: 50})
```

Comment:
No size needed for the dummy I think?

Reply:
I don't know. Do canvas objects with zero size actually get refreshed?

Reply:
There is no longer a need for the dummy - it has been removed.